presto-main/src/main/java/com/facebook/presto/operator/SpatialIndexBuilderOperator.java

```diff
@@ -0,0 +1,207 @@
+                    expectedPositions,
+                    pagesIndexFactory);
+
+            return operator;
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("Spatial index build can not be duplicated");
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final LocalMemoryContext localUserMemoryContext;
+    private final PagesSpatialIndexFactory pagesSpatialIndexFactory;
+
+    private final List<Integer> outputChannels;
+    private final int indexChannel;
+    private final Optional<JoinFilterFunctionFactory> filterFunctionFactory;
+
+    private final PagesIndex index;
+
+    private boolean finished;
+
+    public SpatialIndexBuilderOperator(
+            OperatorContext operatorContext,
+            PagesSpatialIndexFactory pagesSpatialIndexFactory,
+            List<Integer> outputChannels,
+            int indexChannel,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory,
+            int expectedPositions,
+            PagesIndex.Factory pagesIndexFactory)
+    {
+        this.operatorContext = operatorContext;
+        this.localUserMemoryContext = operatorContext.localUserMemoryContext();
+        this.filterFunctionFactory = filterFunctionFactory;
+
+        this.index = pagesIndexFactory.newPagesIndex(pagesSpatialIndexFactory.getTypes(), expectedPositions);
+        this.pagesSpatialIndexFactory = pagesSpatialIndexFactory;
+
+        this.outputChannels = outputChannels;
+        this.indexChannel = indexChannel;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return pagesSpatialIndexFactory.getTypes();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+        checkState(!isFinished(), "Operator is already finished");
+
+        index.addPage(page);
+
+        if (!localUserMemoryContext.trySetBytes((index.getEstimatedSize().toBytes()))) {
+            index.compact();
+            localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes());
+        }
+
+        operatorContext.recordGeneratedOutput(page.getSizeInBytes(), page.getPositionCount());
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    @Override
+    public void finish()
+    {
+        if (finished) {
+            return;
+        }
+
+        PagesSpatialIndex spatialIndex = index.createPagesSpatialIndex(operatorContext.getSession(), indexChannel, filterFunctionFactory, outputChannels);
+        localUserMemoryContext.setBytes(spatialIndex.getEstimatedMemorySizeInBytes());
```

Comment:
@findepi Strictly speaking, `spatialIndex == EMPTY_INDEX` implies that all records in the `index` have their `geometry` column `null` or empty. While I expect that in most cases the number of such records will be very small, it is technically possible for the `index` to have lots of large non-geometry columns and lots of rows with `null` or empty geometries. In this case, index size may be significant. However, since `SpatialIndexBuilderOperator` is finished, won't it become eligible for garbage collection which will free up memory used by the `index`?

My reading of the code suggests that `PagesIndex` and `PagesSpatialIndex` share `valueAddresses` in full and share `channels` mostly (e.g. `channels` is a list-of-lists and inner lists are shares while the outer list is not). Do you agree? If so, then the extra memory used by `PagesIndex` that's not accounted for is memory used by the outer list of `channels`. This is a list of pointers whose size is equal to the number of columns. Looking at `SimplePagesHashStrategy` whose `getSizeInBytes` method doesn't account for the lists of `types`, `outputChannels` and `hashChannels` suggests that it is OK to omit small items like this.